pkg/controller/instance/instance_controller.go

```diff
@@ -19,94 +19,55 @@ import (
+	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
-	"sigs.k8s.io/controller-runtime/pkg/controller"
-	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
-	"sigs.k8s.io/controller-runtime/pkg/event"
-	"sigs.k8s.io/controller-runtime/pkg/handler"
-	"sigs.k8s.io/controller-runtime/pkg/manager"
-	"sigs.k8s.io/controller-runtime/pkg/predicate"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
-	"sigs.k8s.io/controller-runtime/pkg/source"
 )
 
-// Add creates a new Instance Controller and adds it to the Manager with default RBAC.
-//
-// The Manager will set fields on the Controller and start it when the Manager is started.
-func Add(mgr manager.Manager) error {
-	log.Printf("InstanceController: Registering instance controller.")
-	return add(mgr, newReconciler(mgr))
-}
-
-// newReconciler returns a new reconcile.Reconciler
-func newReconciler(mgr manager.Manager) reconcile.Reconciler {
-	return &ReconcileInstance{Client: mgr.GetClient(), scheme: mgr.GetScheme(), recorder: mgr.GetEventRecorderFor("instance-controller")}
+// Reconciler reconciles an Instance object.
+type Reconciler struct {
+	client.Client
+	Recorder record.EventRecorder
+	Scheme   *runtime.Scheme
 }
 
-// add adds a new Controller to mgr with r as the reconcile.Reconciler.
-func add(mgr manager.Manager, r reconcile.Reconciler) error {
-	// Create a new controller
-	c, err := controller.New("instance-controller", mgr, controller.Options{Reconciler: r})
-	if err != nil {
-		return err
-	}
-
-	// Watch for changes to Instance
-	if err = c.Watch(&source.Kind{Type: &kudov1alpha1.Instance{}}, &handler.EnqueueRequestForObject{}, instanceEventFilter(mgr)); err != nil {
-		return err
-	}
-
-	// Watch for changes to OperatorVersion. Since changes to OperatorVersion and Instance are often happening
-	// concurrently there is an inherent race between both update events so that we might see a new Instance first
-	// without the corresponding OperatorVersion. We additionally watch OperatorVersions and trigger
-	// reconciliation for the corresponding instances.
-	//
-	// Define a mapping from the object in the event (OperatorVersion) to one or more objects to
-	// reconcile (Instances). Specifically this calls for a reconciliation of any owned objects.
-	ovEventHandler := handler.ToRequestsFunc(
+// SetupWithManager registers this reconciler with the controller manager
+func (r *Reconciler) SetupWithManager(
+	mgr ctrl.Manager) error {
+	addOvRelatedInstancesToReconcile := handler.ToRequestsFunc(
 		func(a handler.MapObject) []reconcile.Request {
 			requests := make([]reconcile.Request, 0)
-			// We want to query and queue up operators Instances
 			instances := &kudov1alpha1.InstanceList{}
 			// we are listing all instances here, which could come with some performance penalty
 			// a possible optimization is to introduce filtering based on operatorversion (or operator)
```

Comment:
I'm less concerned with the optimization concern (which we should do... we should put a todo: here)
I'm more concerned that when the watch condition is triggered for 1 instance we perform this operation on ALL instances.     Why do we not work solely with the instance associated with the trigger?

We want to handle edge and level...  we need to handle level in a different way IMO... one that isn't driven by events in the system.

Reply:
ok... after further review... I see that we loop thru all instances... but we only operator on the one defined by "a".   my concern isn't valid... I still suggest the `//TODO:`  but it isn't a big deal.

Reply:
This is nothing new in this PR, this was already in before. I don't like TODOs in code, I am happy to file an issue though :) https://github.com/kudobuilder/kudo/issues/860